README.md
# TNLMeans
TNLMeans is an implementation of the NL-means denoising algorithm in an Avisynth filter (by tritical)

### Requirements

This filter requires AviSynth 2.6.0 or AviSynth+ as well as the Visual C++ Redistributable Package for Visual Studio 2015-19.

## Links

Avisynth wiki: http://avisynth.nl/index.php/TNLMeans

Project on github: https://github.com/pinterf/TNLMeans

### Plugin history

Based on tritical's v1.03 (2007)
Additional works by pinterf
- move to github
- move to Visual Studio 2019
- C++ syntax updates
- Update to be an Avisynth v2.6 interface plugin
- x64 build
- Linux/GCC build
- 10-16 bit and Y, planar YUV/RGB video formats support

see CHANGELOG.md or the documentation

Build instructions
------------------
note: plugin is pure c++, there is no SIMD code inside

### Windows Visual Studio MSVC

use IDE

### Windows GCC

(mingw installed by msys2)
 From the 'build' folder under project root:

```
del ..\CMakeCache.txt
cmake .. -G "MinGW Makefiles"
cmake --build . --config Release
```

### Linux

Clone repo and build

```
git clone https://github.com/pinterf/TNLMeans
cd TNLMeans
cmake -B build -S .
cmake --build build
```

Useful hints:

build after clean:

```
cmake --build build --clean-first
```

delete CMake cache:

```
rm build/CMakeCache.txt
```



* Find binaries at
  
        build/TNLMeans/libtnlmeans.so

* Install binaries

        cd build
        sudo make install



Documentation/CHANGELOG.md
# TNLMeans Changelog

**v1.1 (20210410 - pinterf)**

- project moved to github: https://github.com/pinterf/TNLMeans
- project update to Visual Studio 2019
- Change to Avisynth v2.6 interface (from 2.5)
- pass frame properties for AviSynth+ interface version >= 8
- Source code syntax updates: gcc, clang
- Add planar Y, YUV and RGB formats (but kept YUY2 by on-the-fly pre-post conversion)
- Add x64 config
- linux/gcc build, CMake build environment
- 10-16 bit support

**v1.0.3 (20070828  tritical's last version)**

- Removed fast exp() approximation that was used for sse=false.  Turns out it was quite inaccurate and had overflow problems resulting in artifacts.

**v1.0.2 (20060730)** 

- Fixed a problem with small weights causing artifacts

**v1.0.1 (20060619)**

- Fixed a bug that caused a crash when ms=true was used with yuy2 input

**v1.0 Final (20060531)**

- Fixed always creating the downsampled clip unless ms=false was explicitly specified

**v1.0 Beta 2 (20060525)**

- Added multiscale version (parameters ms/rm)
- Added sse parameter (sum of squared errors)
- optimized non-block based routines by buffering (100% speed increase)
- removed b parameter
- fixed a bug in the block based routines that caused some blocks in the search window not to be tested
- changed defaults for ax/ay/sx/sy/h

**v1.0 Beta 1 (20060517 - tritical)**

- Initial Release



TNLMeans/common.h
#ifndef __COMMON_H__
#define __COMMON_H__

#include "avisynth.h"
#include <stdint.h>

#if defined(__clang__)
// Check clang first. clang-cl also defines __MSC_VER
// We set MSVC because they are mostly compatible
#   define CLANG
#if defined(_MSC_VER)
#   define MSVC
#   define TN_FORCEINLINE __attribute__((always_inline))
#else
#   define TN_FORCEINLINE __attribute__((always_inline)) inline
#endif
#elif   defined(_MSC_VER)
#   define MSVC
#   define MSVC_PURE
#   define TN_FORCEINLINE __forceinline
#elif defined(__GNUC__)
#   define GCC
#   define TN_FORCEINLINE __attribute__((always_inline)) inline
#else
#   error Unsupported compiler.
#   define TN_FORCEINLINE inline
#   undef __forceinline
#   define __forceinline inline
#endif 


#ifndef _WIN32
#define OutputDebugString(x)
#endif

#if (defined(GCC) || defined(CLANG)) && !defined(_WIN32)
#include <stdlib.h>
#define _aligned_malloc(size, alignment) aligned_alloc(alignment, size)
#define _aligned_free(ptr) free(ptr)
#endif

#ifndef _WIN32
#include <stdio.h>
#ifdef AVS_POSIX
#ifndef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE 1
#endif
#include <limits.h>
#endif
#endif

#endif


TNLMeans/PlanarFrame.h
/*
**   My PlanarFrame class... fast mmx/sse2 YUY2 packed to planar and planar
**   to packed conversions, and always gives 16 bit alignment for all
**   planes.  Supports YV12/YUY2 frames from avisynth, can do any planar format
**   internally.
**
**   Copyright (C) 2005-2006 Kevin Stone
**
**   This program is free software; you can redistribute it and/or modify
**   it under the terms of the GNU General Public License as published by
**   the Free Software Foundation; either version 2 of the License, or
**   (at your option) any later version.
**
**   This program is distributed in the hope that it will be useful,
**   but WITHOUT ANY WARRANTY; without even the implied warranty of
**   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**   GNU General Public License for more details.
**
**   You should have received a copy of the GNU General Public License
**   along with this program; if not, write to the Free Software
**   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef __PlanarFrame_H__
#define __PlanarFrame_H__

#include "avisynth.h"
#include <stdlib.h>
#include <stdint.h>

#define MIN_ALIGNMENT 64

#define PLANAR_420 1
#define PLANAR_422 2
#define PLANAR_444 3
#define PLANAR_411 4
#define PLANAR_400 5

class PlanarFrame
{
private:
  int cpu;
  bool useSIMD, packed;
  int ypitch, uvpitch;
  int ywidth, uvwidth;
  int yheight, uvheight;
  uint8_t *y, *u, *v;
  bool allocSpace(VideoInfo &viInfo);
  bool allocSpace(int specs[4]);
  void copyInternalFrom(PVideoFrame &frame, VideoInfo &viInfo);
  void copyInternalFrom(PlanarFrame &frame);
  void copyInternalTo(PVideoFrame &frame, VideoInfo &viInfo);
  void copyInternalTo(PlanarFrame &frame);
  void copyInternalPlaneTo(PlanarFrame &frame, int plane);
  void convYUY2to422(const uint8_t *src, uint8_t *py, uint8_t *pu,
    uint8_t *pv, int pitch1, int pitch2Y, int pitch2UV, int width, int height);
  void convYUY2to422_SSE2(const uint8_t *src, uint8_t *py, uint8_t *pu,
    uint8_t *pv, int pitch1, int pitch2Y, int pitch2UV, int width, int height);
  void conv422toYUY2(uint8_t *py, uint8_t *pu, uint8_t *pv,
    uint8_t *dst, int pitch1Y, int pitch1UV, int pitch2, int width, int height);
  void conv422toYUY2_SSE2(uint8_t *py, uint8_t *pu, uint8_t *pv,
    uint8_t *dst, int pitch1Y, int pitch1UV, int pitch2, int width, int height);

public:
  PlanarFrame(int cpuInfo);
  PlanarFrame(VideoInfo &viInfo, int cpuInfo);
  PlanarFrame(VideoInfo &viInfo, bool _packed, int cpuInfo);
  ~PlanarFrame();
  void createPlanar(int yheight, int uvheight, int ywidth, int uvwidth);
  void createPlanar(int height, int width, int chroma_format);
  void createFromProfile(VideoInfo &viInfo);
  void createFromFrame(PVideoFrame &frame, VideoInfo &viInfo);
  void createFromPlanar(PlanarFrame &frame);
  void copyFrom(PVideoFrame &frame, VideoInfo &viInfo);
  void copyTo(PVideoFrame &frame, VideoInfo &viInfo);
  void copyFrom(PlanarFrame &frame);
  void copyTo(PlanarFrame &frame);
  void copyChromaTo(PlanarFrame &dst);
  void copyToForBMP(PVideoFrame &dst, VideoInfo &viInfo);
  void copyPlaneTo(PlanarFrame &dst, int plane);
  void freePlanar();
  uint8_t* GetPtr(int plane = 0);
  int NumComponents();
  int GetWidth(int plane = 0);
  int GetHeight(int plane = 0);
  int GetPitch(int plane = 0);
  void BitBlt(uint8_t* dstp, int dst_pitch, const uint8_t* srcp,
    int src_pitch, int row_size, int height);
  PlanarFrame& operator=(PlanarFrame &ob2);
};

#endif


TNLMeans/PlanarFrame.cpp
/*
**   My PlanarFrame class... fast mmx/sse2 YUY2 packed to planar and planar
**   to packed conversions, and always gives 16 bit alignment for all
**   planes.  Supports YV12/YUY2 frames from avisynth, can do any planar format
**   internally.
**
**   Copyright (C) 2005-2006 Kevin Stone
**
**   This program is free software; you can redistribute it and/or modify
**   it under the terms of the GNU General Public License as published by
**   the Free Software Foundation; either version 2 of the License, or
**   (at your option) any later version.
**
**   This program is distributed in the hope that it will be useful,
**   but WITHOUT ANY WARRANTY; without even the implied warranty of
**   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**   GNU General Public License for more details.
**
**   You should have received a copy of the GNU General Public License
**   along with this program; if not, write to the Free Software
**   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "common.h"
#include "PlanarFrame.h"
#include "avs/cpuid.h"
#include <emmintrin.h>
#include <stdint.h>

// 8 bits only!!!

PlanarFrame::PlanarFrame(int cpuFlags)
{
  ypitch = uvpitch = 0;
  ywidth = uvwidth = 0;
  yheight = uvheight = 0;
  y = u = v = NULL;
  useSIMD = true;
  packed = false;
  cpu = cpuFlags;
}

PlanarFrame::PlanarFrame(VideoInfo &viInfo, int cpuFlags)
{
  ypitch = uvpitch = 0;
  ywidth = uvwidth = 0;
  yheight = uvheight = 0;
  y = u = v = NULL;
  useSIMD = true;
  packed = false;
  cpu = cpuFlags;
  allocSpace(viInfo);
}

PlanarFrame::PlanarFrame(VideoInfo &viInfo, bool _packed, int cpuFlags)
{
  ypitch = uvpitch = 0;
  ywidth = uvwidth = 0;
  yheight = uvheight = 0;
  y = u = v = NULL;
  useSIMD = true;
  packed = _packed;
  cpu = cpuFlags;
  allocSpace(viInfo);
}

PlanarFrame::~PlanarFrame()
{
  if (y != NULL) { _aligned_free(y); y = NULL; }
  if (u != NULL) { _aligned_free(u); u = NULL; }
  if (v != NULL) { _aligned_free(v); v = NULL; }
}

bool PlanarFrame::allocSpace(VideoInfo &viInfo)
{
  if (y != NULL) { _aligned_free(y); y = NULL; }
  if (u != NULL) { _aligned_free(u); u = NULL; }
  if (v != NULL) { _aligned_free(v); v = NULL; }
  int height = viInfo.height;
  int width = viInfo.width;
  if (viInfo.IsPlanar())
  {
    ypitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
    ywidth = width;
    yheight = height;
    y = (uint8_t*)_aligned_malloc(ypitch * yheight, MIN_ALIGNMENT);
    if (y == NULL) return false;

    if (!viInfo.IsY()) {
      width >>= viInfo.GetPlaneWidthSubsampling(PLANAR_U);
      height >>= viInfo.GetPlaneHeightSubsampling(PLANAR_U);
      uvpitch = width + ((width % MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width % MIN_ALIGNMENT));
      uvwidth = width;
      uvheight = height;
      u = (uint8_t*)_aligned_malloc(uvpitch * uvheight, MIN_ALIGNMENT);
      if (u == NULL) return false;
      v = (uint8_t*)_aligned_malloc(uvpitch * uvheight, MIN_ALIGNMENT);
      if (v == NULL) return false;
    }
    return true;
  }
  else if (viInfo.IsYUY2())
  {
    if (!packed)
    {
      ypitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
      ywidth = width;
      yheight = height;
      width >>= 1;
      uvpitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
      uvwidth = width;
      uvheight = height;
      y = (uint8_t*)_aligned_malloc(ypitch*yheight, MIN_ALIGNMENT);
      if (y == NULL) return false;
      u = (uint8_t*)_aligned_malloc(uvpitch*uvheight, MIN_ALIGNMENT);
      if (u == NULL) return false;
      v = (uint8_t*)_aligned_malloc(uvpitch*uvheight, MIN_ALIGNMENT);
      if (v == NULL) return false;
      return true;
    }
    else
    {
      width *= 2;
      ypitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
      ywidth = width;
      yheight = height;
      y = (uint8_t*)_aligned_malloc(ypitch*yheight, MIN_ALIGNMENT);
      if (y == NULL) return false;
      uvpitch = uvwidth = uvheight = 0;
      u = v = NULL;
      return true;
    }
  }
  return false;
}

bool PlanarFrame::allocSpace(int specs[4])
{
  if (y != NULL) { _aligned_free(y); y = NULL; }
  if (u != NULL) { _aligned_free(u); u = NULL; }
  if (v != NULL) { _aligned_free(v); v = NULL; }
  int height = specs[0];
  int width = specs[2];
  ypitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
  ywidth = width;
  yheight = height;
  height = specs[1];
  width = specs[3];
  uvpitch = width + ((width%MIN_ALIGNMENT) == 0 ? 0 : MIN_ALIGNMENT - (width%MIN_ALIGNMENT));
  uvwidth = width;
  uvheight = height;

  y = (uint8_t*)_aligned_malloc(ypitch*yheight, MIN_ALIGNMENT);
  if (y == NULL) return false;
  if (uvpitch) {
    u = (uint8_t*)_aligned_malloc(uvpitch * uvheight, MIN_ALIGNMENT);
    if (u == NULL) return false;
    v = (uint8_t*)_aligned_malloc(uvpitch * uvheight, MIN_ALIGNMENT);
    if (v == NULL) return false;
  }
  return true;
}


void PlanarFrame::createPlanar(int yheight, int uvheight, int ywidth, int uvwidth)
{
  int specs[4] = { yheight, uvheight, ywidth, uvwidth };
  allocSpace(specs);
}

void PlanarFrame::createPlanar(int height, int width, int chroma_format)
{
  int specs[4];
  if (chroma_format <= PLANAR_420) // 420
  {
    specs[0] = height; specs[1] = height >> 1;
    specs[2] = width; specs[3] = width >> 1;
  }
  else if (chroma_format == PLANAR_422) // 422
  {
    specs[0] = height; specs[1] = height;
    specs[2] = width; specs[3] = width >> 1;
  }
  else if (chroma_format == PLANAR_444) // 444
  {
    specs[0] = height; specs[1] = height;
    specs[2] = width; specs[3] = width;
  }
  else if (chroma_format == PLANAR_411) // 411
  {
    specs[0] = height; specs[1] = height;
    specs[2] = width; specs[3] = width >> 2;
  }
  else if (chroma_format == PLANAR_400) // greyscale
  {
    specs[0] = height; specs[1] = 0;
    specs[2] = width; specs[3] = 0;
  }
  allocSpace(specs);
}

void PlanarFrame::createFromProfile(VideoInfo &viInfo)
{
  allocSpace(viInfo);
}

void PlanarFrame::createFromFrame(PVideoFrame &frame, VideoInfo &viInfo)
{
  allocSpace(viInfo);
  copyInternalFrom(frame, viInfo);
}

void PlanarFrame::createFromPlanar(PlanarFrame &frame)
{
  int specs[4] = { frame.yheight, frame.uvheight, frame.ywidth, frame.uvwidth };
  allocSpace(specs);
  copyInternalFrom(frame);
}

void PlanarFrame::copyFrom(PVideoFrame &frame, VideoInfo &viInfo)
{
  copyInternalFrom(frame, viInfo);
}

void PlanarFrame::copyFrom(PlanarFrame &frame)
{
  copyInternalFrom(frame);
}

void PlanarFrame::copyTo(PVideoFrame &frame, VideoInfo &viInfo)
{
  copyInternalTo(frame, viInfo);
}

void PlanarFrame::copyTo(PlanarFrame &frame)
{
  copyInternalTo(frame);
}

void PlanarFrame::copyPlaneTo(PlanarFrame &frame, int plane)
{
  copyInternalPlaneTo(frame, plane);
}

uint8_t* PlanarFrame::GetPtr(int plane)
{
  if (plane == 0) return y;
  if (plane == 1) return u;
  return v;
}

int PlanarFrame::NumComponents() {
  if (uvpitch)
    return 3;
  return 1;
}

int PlanarFrame::GetWidth(int plane)
{
  if (plane == 0) return ywidth;
  else return uvwidth;
}

int PlanarFrame::GetHeight(int plane)
{
  if (plane == 0) return yheight;
  else return uvheight;
}

int PlanarFrame::GetPitch(int plane)
{
  if (plane == 0) return ypitch;
  else return uvpitch;
}

void PlanarFrame::freePlanar()
{
  if (y != NULL) { _aligned_free(y); y = NULL; }
  if (u != NULL) { _aligned_free(u); u = NULL; }
  if (v != NULL) { _aligned_free(v); v = NULL; }
  ypitch = uvpitch = 0;
  ywidth = uvwidth = 0;
  yheight = uvheight = 0;
}

void PlanarFrame::copyInternalFrom(PVideoFrame &frame, VideoInfo &viInfo)
{
  if (y == NULL) return;
  if (viInfo.IsPlanar())
  {
    BitBlt(y, ypitch, frame->GetReadPtr(PLANAR_Y), frame->GetPitch(PLANAR_Y),
      frame->GetRowSize(PLANAR_Y), frame->GetHeight(PLANAR_Y));
    if (u == NULL || v == NULL) return;
    BitBlt(u, uvpitch, frame->GetReadPtr(PLANAR_U), frame->GetPitch(PLANAR_U),
      frame->GetRowSize(PLANAR_U), frame->GetHeight(PLANAR_U));
    BitBlt(v, uvpitch, frame->GetReadPtr(PLANAR_V), frame->GetPitch(PLANAR_V),
      frame->GetRowSize(PLANAR_V), frame->GetHeight(PLANAR_V));
  }
  else if (viInfo.IsYUY2())
  {
    convYUY2to422(frame->GetReadPtr(), y, u, v, frame->GetPitch(), ypitch, uvpitch,
      viInfo.width, viInfo.height);
  }
}

void PlanarFrame::copyInternalFrom(PlanarFrame &frame)
{
  if (y == NULL) return;
  BitBlt(y, ypitch, frame.y, frame.ypitch, frame.ywidth, frame.yheight);
  if (u == NULL || v == NULL) return;
  BitBlt(u, uvpitch, frame.u, frame.uvpitch, frame.uvwidth, frame.uvheight);
  BitBlt(v, uvpitch, frame.v, frame.uvpitch, frame.uvwidth, frame.uvheight);
}

void PlanarFrame::copyInternalTo(PVideoFrame &frame, VideoInfo &viInfo)
{
  if (y == NULL) return;
  if (viInfo.IsPlanar())
  {
    BitBlt(frame->GetWritePtr(PLANAR_Y), frame->GetPitch(PLANAR_Y), y, ypitch, ywidth, yheight);
    if (u == NULL || v == NULL) return;
    BitBlt(frame->GetWritePtr(PLANAR_U), frame->GetPitch(PLANAR_U), u, uvpitch, uvwidth, uvheight);
    BitBlt(frame->GetWritePtr(PLANAR_V), frame->GetPitch(PLANAR_V), v, uvpitch, uvwidth, uvheight);
  }
  else if (viInfo.IsYUY2())
  {
    conv422toYUY2(y, u, v, frame->GetWritePtr(), ypitch, uvpitch, frame->GetPitch(), ywidth, yheight);
  }
}

void PlanarFrame::copyInternalTo(PlanarFrame &frame)
{
  if (y == NULL) return;
  BitBlt(frame.y, frame.ypitch, y, ypitch, ywidth, yheight);
  if (u == NULL || v == NULL) return;
  BitBlt(frame.u, frame.uvpitch, u, uvpitch, uvwidth, uvheight);
  BitBlt(frame.v, frame.uvpitch, v, uvpitch, uvwidth, uvheight);
}

void PlanarFrame::copyInternalPlaneTo(PlanarFrame &frame, int plane)
{
  if (plane == 0 && y != NULL)
    BitBlt(frame.y, frame.ypitch, y, ypitch, ywidth, yheight);
  else if (plane == 1 && u != NULL)
    BitBlt(frame.u, frame.uvpitch, u, uvpitch, uvwidth, uvheight);
  else if (plane == 2 && v != NULL)
    BitBlt(frame.v, frame.uvpitch, v, uvpitch, uvwidth, uvheight);
}

void PlanarFrame::copyChromaTo(PlanarFrame &dst)
{
  BitBlt(dst.u, dst.uvpitch, u, uvpitch, dst.uvwidth, dst.uvheight);
  BitBlt(dst.v, dst.uvpitch, v, uvpitch, dst.uvwidth, dst.uvheight);
}

void PlanarFrame::copyToForBMP(PVideoFrame &dst, VideoInfo &viInfo)
{
  uint8_t *dstp = dst->GetWritePtr(PLANAR_Y);
  if (viInfo.IsPlanar())
  {
    int out_pitch = (ywidth + 3) & -4;
    BitBlt(dstp, out_pitch, y, ypitch, ywidth, yheight);
    BitBlt(dstp + (out_pitch*yheight), out_pitch >> 1, v, uvpitch, uvwidth, uvheight);
    BitBlt(dstp + (out_pitch*yheight) + ((out_pitch >> 1)*uvheight), out_pitch >> 1, u, uvpitch, uvwidth, uvheight);
  }
  else
  {
    int out_pitch = (dst->GetRowSize(PLANAR_Y) + 3) & -4;
    conv422toYUY2(y, u, v, dstp, ypitch, uvpitch, out_pitch, viInfo.width, viInfo.height);
  }
}

PlanarFrame& PlanarFrame::operator=(PlanarFrame &ob2)
{
  cpu = ob2.cpu;
  ypitch = ob2.ypitch;
  yheight = ob2.yheight;
  ywidth = ob2.ywidth;
  uvpitch = ob2.uvpitch;
  uvheight = ob2.uvheight;
  uvwidth = ob2.uvwidth;
  this->copyFrom(ob2);
  return *this;
}

void PlanarFrame::convYUY2to422(const uint8_t *src, uint8_t *py, uint8_t *pu,
  uint8_t *pv, int pitch1, int pitch2Y, int pitch2UV, int width, int height)
{
  if ((cpu&CPUF_SSE2) && useSIMD)
    convYUY2to422_SSE2(src, py, pu, pv, pitch1, pitch2Y, pitch2UV, width, height);
  else
  {
    width >>= 1;
    for (int y = 0; y < height; ++y)
    {
      for (int x = 0; x < width; ++x)
      {
        py[x << 1] = src[x << 2];
        pu[x] = src[(x << 2) + 1];
        py[(x << 1) + 1] = src[(x << 2) + 2];
        pv[x] = src[(x << 2) + 3];
      }
      py += pitch2Y;
      pu += pitch2UV;
      pv += pitch2UV;
      src += pitch1;
    }
  }
}


void PlanarFrame::convYUY2to422_SSE2(const uint8_t *src, uint8_t *py, uint8_t *pu,
  uint8_t *pv, int pitch1, int pitch2Y, int pitch2UV, int width, int height)
{
  width >>= 1; // mov ecx, width
  __m128i Ymask = _mm_set1_epi16(0x00FF);
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x += 4) {
      __m128i fullsrc = _mm_load_si128(reinterpret_cast<const __m128i *>(src + x * 4)); // VYUYVYUYVYUYVYUY
      __m128i yy = _mm_and_si128(fullsrc, Ymask); // 0Y0Y0Y0Y0Y0Y0Y0Y
      __m128i uvuv = _mm_srli_epi16(fullsrc, 8); // 0V0U0V0U0V0U0V0U
      yy = _mm_packus_epi16(yy, yy); // xxxxxxxxYYYYYYYY
      uvuv = _mm_packus_epi16(uvuv, uvuv); // xxxxxxxxVUVUVUVU
      __m128i uu = _mm_and_si128(uvuv, Ymask); // xxxxxxxx0U0U0U0U
      __m128i vv = _mm_srli_epi16(uvuv, 8); // xxxxxxxx0V0V0V0V
      uu = _mm_packus_epi16(uu, uu); // xxxxxxxxxxxxUUUU
      vv = _mm_packus_epi16(vv, vv); // xxxxxxxxxxxxVVVV
      _mm_storel_epi64(reinterpret_cast<__m128i *>(py + x * 2), yy); // store y
      *(uint32_t *)(pu + x) = _mm_cvtsi128_si32(uu); // store u
      *(uint32_t *)(pv + x) = _mm_cvtsi128_si32(vv); // store v
    }
    src += pitch1;
    py += pitch2Y;
    pu += pitch2UV;
    pv += pitch2UV;
  }
}

void PlanarFrame::conv422toYUY2(uint8_t *py, uint8_t *pu, uint8_t *pv,
  uint8_t *dst, int pitch1Y, int pitch1UV, int pitch2, int width, int height)
{
  if ((cpu&CPUF_SSE2) && useSIMD)
    conv422toYUY2_SSE2(py, pu, pv, dst, pitch1Y, pitch1UV, pitch2, width, height);
  else
  {
    width >>= 1;
    for (int y = 0; y < height; ++y)
    {
      for (int x = 0; x < width; ++x)
      {
        dst[x << 2] = py[x << 1];
        dst[(x << 2) + 1] = pu[x];
        dst[(x << 2) + 2] = py[(x << 1) + 1];
        dst[(x << 2) + 3] = pv[x];
      }
      py += pitch1Y;
      pu += pitch1UV;
      pv += pitch1UV;
      dst += pitch2;
    }
  }
}


void PlanarFrame::conv422toYUY2_SSE2(uint8_t *py, uint8_t *pu, uint8_t *pv,
  uint8_t *dst, int pitch1Y, int pitch1UV, int pitch2, int width, int height)
{
  width >>= 1; // mov ecx, width
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x += 4) {
      __m128i yy = _mm_loadl_epi64(reinterpret_cast<const __m128i *>(py + x * 2)); // YYYYYYYY
      __m128i uu = _mm_castps_si128(_mm_load_ss(reinterpret_cast<float *>(pu + x))); // 000000000000UUUU
      __m128i vv = _mm_castps_si128(_mm_load_ss(reinterpret_cast<float *>(pv + x))); // 000000000000VVVV
      __m128i uvuv = _mm_unpacklo_epi8(uu, vv); // 00000000VUVUVUVU
      __m128i yuyv = _mm_unpacklo_epi8(yy,uvuv); // VYUYVYUYVYUYVYUY
      _mm_store_si128(reinterpret_cast<__m128i *>(dst + x * 4), yuyv);
    }
    dst += pitch2;
    py += pitch1Y;
    pu += pitch1UV;
    pv += pitch1UV;
  }
}

// Avisynth v2.5.  Copyright 2002 Ben Rudiak-Gould et al.
// http://www.avisynth.org

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA, or visit
// http://www.gnu.org/copyleft/gpl.html .
//
// Linking Avisynth statically or dynamically with other modules is making a
// combined work based on Avisynth.  Thus, the terms and conditions of the GNU
// General Public License cover the whole combination.
//
// As a special exception, the copyright holders of Avisynth give you
// permission to link Avisynth with independent modules that communicate with
// Avisynth solely through the interfaces defined in avisynth.h, regardless of the license
// terms of these independent modules, and to copy and distribute the
// resulting combined work under terms of your choice, provided that
// every copy of the combined work is accompanied by a complete copy of
// the source code of Avisynth (the version of Avisynth used to produce the
// combined work), being distributed under the terms of the GNU General
// Public License plus this exception.  An independent module is a module
// which is not derived from or based on Avisynth, such as 3rd-party filters,
// import and export plugins, or graphical user interfaces.

// from Avisynth 2.55 source...
// copied so we don't need an
// IScriptEnvironment pointer 
// to call it

#include "avisynth.h"
#include <memory.h>

void PlanarFrame::BitBlt(uint8_t* dstp, int dst_pitch, const uint8_t* srcp,
  int src_pitch, int row_size, int height)
{
  if (!height || !row_size) return;
  if (height == 1 || (dst_pitch == src_pitch && src_pitch == row_size))
    memcpy(dstp, srcp, src_pitch * height);
  else
  {
    for (int y = height; y > 0; --y)
    {
      memcpy(dstp, srcp, row_size);
      dstp += dst_pitch;
      srcp += src_pitch;
    }
  }
}

